thirdparty/kyaml/runfn/runfn.go

```diff
@@ -152,37 +131,86 @@ func (r RunFns) getNodesAndFilters() (
 		return nil, nil, outputPkg, err
 	}
 
-	fltrs, err := r.getFilters(buff.Nodes)
+	fltrs, err := r.getFilters()
 	if err != nil {
 		return nil, nil, outputPkg, err
 	}
+
+	// we need to exclude function config file if we don't want to include
+	// meta resources.
+	if !r.IncludeMetaResources {
+		p := kio.Pipeline{
+			Inputs:                []kio.Reader{buff},
+			Filters:               []kio.Filter{kio.FilterFunc(r.filterOutFnConfigFile)},
+			Outputs:               []kio.Writer{buff},
+			ContinueOnEmptyResult: r.ContinueOnEmptyResult,
+		}
+		if err := p.Execute(); err != nil {
+			return nil, nil, outputPkg, err
+		}
+	}
 	return buff, fltrs, outputPkg, nil
 }
 
-func (r RunFns) getFilters(nodes []*yaml.RNode) ([]kio.Filter, error) {
-	var fltrs []kio.Filter
-
-	// fns from annotations on the input resources
-	f, err := r.getFunctionsFromInput(nodes)
-	if err != nil {
-		return nil, err
+// filterOutFnConfigFile is a filter which filters out the function config
+// file
+func (r RunFns) filterOutFnConfigFile(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	var result []*yaml.RNode
+	// fn config path should be absolute
+	var fnConfigPath string
+	if filepath.IsAbs(r.FnConfigPath) {
+		fnConfigPath = r.FnConfigPath
+	} else {
+		fnConfigPath = filepath.Join(r.Path, r.FnConfigPath)
 	}
-	fltrs = append(fltrs, f...)
+	for i := range nodes {
+		node := nodes[i]
+		meta, err := node.GetMeta()
+		if err != nil {
+			return nil, err
+		}
+		path := meta.Annotations[kioutil.PathAnnotation]
+		if path != "" {
+			// file path in the annotation is relative to the package path
+			// we need to use absolute path
+			path = filepath.Join(r.Path, path)
```

Comment:
you are conflating OS paths and OS-agnostics paths here.

@droot Same discussion as in packaging PRs.

Reply:
I followed the same logic used in kio to write back the files here.

Reply:
I think this is not conflating. The `path` read from annotations is added in `kio.LocalPackageReader` by `filepath` so it's an OS specific path. `r.Path` is read from command line arguments so it's OS specific as well.

Reply:
The code is targeting OS-specific paths here in this case. Not sure what @frankfarzan is referring to.

Reply:
It's entirely possible this is an existing bug.

These annotations are consumed by functions and need to be OS-agnostic:

https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/config-io.md#configkubernetesiopath

AI:

1. This also highlights the need for extracting a common package for dealing with these annotations (path, index, package). You shouldn't be doing this low-level operation here.
2. Need to add unit tests that would catch this bug.

Reply:
Yes, according to the conventions the implementation in kyaml is not correct. But that issue only occurs when

 1. Function container and host system are using different path format.
 2. Function needs to read and use the path in the annotations.

I believe so far we haven't had any bug reported about this in both kpt and kustomize. And I agree that this operations should be implemented in a separate package.

Reply:
Yes, probably the easiest way to hit this bug is to run `kpt fn run --exec` on windows.

Is this fix in this PR or a follow up?

Reply:
I don't think exec function on windows can lead to this bug since all paths are windows-style. The only way to reproduce this bug is using linux container on windows and I cannot see a reason for the function to read and use the path stored in annotation in this case.

Reply:
You have to assume the function will treat the annotation as OS-agnostic slash seperated path. It's part of the API (see the link above). So if the orchestrator passes a OS-specific path, you'll hit this bug. Actually whether the function is containerized or not doesn't matter, what matters is the orchestrator is running on a non-POSIX system.

In any case, this is an obvious bug that needs to be fixed.

Reply:
I see. I think this bug should not only be fixed here and also needs to be fixed in kyaml so it's not covered by this PR.

Reply:
And probably kyaml should provide a bunch of methods to process these paths.

Reply:
Add a TODO for this.

Reply:
Had a discussion with Donny about this. I am suggesting to remove the filtering of function config from this PR. Here are couple of more thoughts we discussed:

Imagine a user having multiple function config files that they use for different functions they invoke imperatively. This exclusion of single file does not solve the real problem. Instead excluding using the `krmignore` file will be right pattern for it.

Another thinking is the right way to add this filtering will be to extend `ignoreFile` filter in `kio` to enable this functionality.

Reply:
As discussed offline, the function config file for `kpt fn eval` should not be in the package. If it is, it will be processed as a normal KRM resource. If users want to avoid this they need to add the files to `.krmignore`. But we don't recommend it and will not document it as well.

Reply:
Create an issue in kyaml for the annotation path format problem.

https://github.com/kubernetes-sigs/kustomize/issues/3749